File: CS/Form1.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Linq;
using System.Text;
using System.Windows.Forms;
using System.Collections;
using DevExpress.XtraGrid.Views.Grid;

namespace WindowsFormsApplication186
{
    public partial class Form1 : Form
    {
        public Form1()
        {
            InitializeComponent();
        }

        Hashtable ht = new Hashtable();
        List<DataRow> rowsInProcess = new List<DataRow>();
        private void Form1_Load(object sender, EventArgs e)
        {
            for(int i =0; i<5; i++)
                dataTable1.Rows.Add(new object[] { i, "Name" + i});

        }

        private void repositoryItemButtonEdit1_ButtonClick(object sender, DevExpress.XtraEditors.Controls.ButtonPressedEventArgs e)
        {
            Timer t = new Timer();
            t.Interval = 2000;
            t.Tick += new EventHandler(t_Tick);
            GridView view = (GridView)gridControl1.FocusedView;
            DataRow dr = view.GetDataRow(view.FocusedRowHandle);
            view.CloseEditor();
            view.UpdateCurrentRow();
            t.Tag = dr;
            t.Start();
            rowsInProcess.Add(dr);
        }

        void t_Tick(object sender, EventArgs e)
        {
            Timer t = (Timer)sender;
            DataRow dr = t.Tag as DataRow;
            rowsInProcess.Remove(dr);
            gridView1.CloseEditor();
            gridView1.RefreshRow(FindRowHandleByDataRow(gridView1, dr));
            t.Stop();
        }

        private void gridView1_CustomRowCellEdit(object sender, CustomRowCellEditEventArgs e)
        {
            if (e.Column.FieldName == "Name") {
                GridView view = (GridView)sender;
                DataRow dr = view.GetDataRow(e.RowHandle);
                if (rowsInProcess.Contains(dr))
                    e.RepositoryItem = riDisabledButtonEdit;
                else
                    e.RepositoryItem = riEnabledButtonEdit;
            }
        }


        private int FindRowHandleByDataRow(DevExpress.XtraGrid.Views.Grid.GridView view, DataRow row)
        {
            for (int i = 0; i < view.DataRowCount; i++)
                if (view.GetDataRow(i) == row)
                    return i;
            return DevExpress.XtraGrid.GridControl.InvalidRowHandle;
        }
    }
}

File: VB/Form1.vb
Imports Microsoft.VisualBasic
Imports System
Imports System.Collections.Generic
Imports System.ComponentModel
Imports System.Data
Imports System.Drawing
Imports System.Linq
Imports System.Text
Imports System.Windows.Forms
Imports System.Collections
Imports DevExpress.XtraGrid.Views.Grid

Namespace WindowsFormsApplication186
	Partial Public Class Form1
		Inherits Form
		Public Sub New()
			InitializeComponent()
		End Sub

		Private ht As New Hashtable()
		Private rowsInProcess As List(Of DataRow) = New List(Of DataRow)()
		Private Sub Form1_Load(ByVal sender As Object, ByVal e As EventArgs) Handles MyBase.Load
			For i As Integer = 0 To 4
				dataTable1.Rows.Add(New Object() { i, "Name" & i})
			Next i

		End Sub

		Private Sub repositoryItemButtonEdit1_ButtonClick(ByVal sender As Object, ByVal e As DevExpress.XtraEditors.Controls.ButtonPressedEventArgs) Handles riEnabledButtonEdit.ButtonClick
			Dim t As New Timer()
			t.Interval = 2000
			AddHandler t.Tick, AddressOf t_Tick
			Dim view As GridView = CType(gridControl1.FocusedView, GridView)
			Dim dr As DataRow = view.GetDataRow(view.FocusedRowHandle)
			view.CloseEditor()
			view.UpdateCurrentRow()
			t.Tag = dr
			t.Start()
			rowsInProcess.Add(dr)
		End Sub

		Private Sub t_Tick(ByVal sender As Object, ByVal e As EventArgs)
			Dim t As Timer = CType(sender, Timer)
			Dim dr As DataRow = TryCast(t.Tag, DataRow)
			rowsInProcess.Remove(dr)
			gridView1.CloseEditor()
			gridView1.RefreshRow(FindRowHandleByDataRow(gridView1, dr))
			t.Stop()
		End Sub

		Private Sub gridView1_CustomRowCellEdit(ByVal sender As Object, ByVal e As CustomRowCellEditEventArgs) Handles gridView1.CustomRowCellEdit
			If e.Column.FieldName = "Name" Then
				Dim view As GridView = CType(sender, GridView)
				Dim dr As DataRow = view.GetDataRow(e.RowHandle)
				If rowsInProcess.Contains(dr) Then
					e.RepositoryItem = riDisabledButtonEdit
				Else
					e.RepositoryItem = riEnabledButtonEdit
				End If
			End If
		End Sub


		Private Function FindRowHandleByDataRow(ByVal view As DevExpress.XtraGrid.Views.Grid.GridView, ByVal row As DataRow) As Integer
    Dim I As Integer
    For I = 0 To view.DataRowCount - 1
        If row.Equals(view.GetRow(I)) Then
            Return I
        End If
    Next
    Return DevExpress.XtraGrid.GridControl.InvalidRowHandle
		End Function
	End Class
End Namespace
